plutus-core/plutus-ir/test/TypecheckSpec.hs

```diff
@@ -0,0 +1,141 @@
+import PlutusIR.Analysis.Builtins
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Provenance (original)
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusIR.Transform.Beta (beta)
+import PlutusIR.Transform.CaseReduce (caseReduce)
+import PlutusIR.Transform.CommuteFnWithConst (commuteFnWithConst)
+import PlutusIR.Transform.EvaluateBuiltins (evaluateBuiltins)
+import PlutusIR.Transform.Inline.Inline
+import PlutusIR.Transform.StrictifyBindings
+import PlutusIR.Transform.Unwrap (unwrapCancel)
+import PlutusIR.TypeCheck
+import PlutusPrelude (def)
+import Test.QuickCheck (Property, withMaxSuccess)
+import Test.Tasty (testGroup)
+import Test.Tasty.ExpectedFailure
+import Test.Tasty.Extras (TestNested)
+import Test.Tasty.QuickCheck (testProperty)
+
+-- | Typechecking property tests for all PIR compiler passes.
+-- The argument allows the caller to scale the number of tests.
+-- The default for the argument is @1@.
+typecheck_test ::
+  Int -> TestNested
+typecheck_test factor = return $ testGroup "typechecking"
+  [
+  -- pure passes
+    testProperty "unwrap pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure unwrapCancel)
+  , testProperty "caseReduce pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure caseReduce)
+  , testProperty "beta pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure beta)
+  , testProperty "commuteFnWithConst pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure commuteFnWithConst)
+  -- non-pure passes
+  -- FIXME
+  , expectFail $ testProperty "the whole simplifier pass" $
+      withMaxSuccess (factor*3000) prop_typecheck
+  , testProperty "strictifyBindings (Builtin Variant 1) pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_strictifyBindings DefaultFunSemanticsVariant1)
+  , testProperty "strictifyBindings (Builtin Variant 2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_strictifyBindings DefaultFunSemanticsVariant2)
+  , testProperty "evaluateBuiltins (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant1)
+  , testProperty "evaluateBuiltins (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant2)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant1)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant2)
+  ]
+
+-- Convert Either Error () to Either String () to match with the Testable (Either String ())
+-- instance.
+convertToEitherString :: Either (Error DefaultUni DefaultFun ()) ()
+  -> Either String ()
+convertToEitherString = \case
+  Left err -> Left $ show err
+  Right () -> Right ()
+
+-- | Typechecking a term after one of the pure passes.
```

Comment:
This is actually doing more than that! This is not just checking that the term typechecks afterwards, it is checking that it has the same type as it had beforehand. Worth mentioning.